src/Proto3/Wire/Reverse/Internal.hs

```diff
@@ -0,0 +1,717 @@
+-- reachable during builder execution, though completed buffers
+-- may be copied to new storage at any time.  Aside from those
+-- primitives, 'BuildR' implementations may ignore that issue.
+newtype BuildR = BuildR
+  (Addr# -> Int# -> State# RealWorld -> (# Addr#, Int#, State# RealWorld #))
+    -- ^ Both the builder arguments and the returned values are:
+    --
+    --   1. The starting address of the *used* portion of the current buffer.
+    --
+    --   2. The number of *unused* bytes in the current buffer.
+    --
+    --   3. The state token (which does not consume any machine registers).
+    --
+    -- It seems we cannot preserve register allocation between the arguments
+    -- and the returned components, even by including padding.  If GHC were to
+    -- allocate registers right-to-left (instead of the current left-to-right),
+    -- and if it made sure to allocate the register that it uses for closure
+    -- arguments *last* when allocating return registers, then we would stand
+    -- a chance of not having to move the state components between registers.
+    -- That way @a -> b -> 'BuildR'@ and 'BuildR' would use the same registers
+    -- for state components as each other, and a non-inline return from one
+    -- could be used to call the other without moving state components.
+    --
+    -- But in many cases register movements combine with increments.
+    -- Also, we have arranged to put only the most frequently-used state
+    -- components into registers, which reduces the costs of both moves
+    -- and of save/reload pairs.  For example, our tracking of the total
+    -- bytes written involves metadata at the start of the current buffer
+    -- rather than an additional state register.
+
+instance Semigroup BuildR
+  where
+    (<>) = appendBuildR
+    {-# INLINE (<>) #-}
+
+instance Monoid BuildR
+  where
+    mempty = BuildR (\v u s -> (# v, u, s #))
+    {-# INLINE mempty #-}
+
+    mappend = (<>)
+    {-# INLINE mappend #-}
+
+instance Show BuildR
+  where
+    showsPrec prec builder =
+        showParen (prec > 10)
+          (showString "Proto3.Wire.Reverse.lazyByteString " . shows bytes)
+      where
+        bytes = snd (runBuildR builder)
+
+-- | Needed for rewrite rules; normally you would use '<>'.
+appendBuildR  :: BuildR -> BuildR -> BuildR
+appendBuildR = \b c ->
+  let BuildR f = b
+      BuildR g = c
+  in
+    BuildR (\v0 u0 s0 -> case g v0 u0 s0 of (# v1, u1, s1 #) -> f v1 u1 s1)
+{-# INLINE CONLIKE [1] appendBuildR #-}
+
+toBuildR :: (Ptr Word8 -> Int -> IO (Ptr Word8, Int)) -> BuildR
+toBuildR f =
+  BuildR $ \v0 u0 s0 ->
+    let IO g = f (Ptr v0) (I# u0) in
+    case g s0 of (# s1, (Ptr v1, I# u1) #) -> (# v1, u1, s1 #)
+
+fromBuildR :: BuildR -> Ptr Word8 -> Int -> IO (Ptr Word8, Int)
+fromBuildR (BuildR f) (Ptr v0) (I# u0) =
+  IO $ \s0 -> case f v0 u0 s0 of (# v1, u1, s1 #) -> (# s1, (Ptr v1, I# u1) #)
+
+-- | Forces eta expansion of a function that produces a 'BuildR', so that
```

Comment:
Is this function used for performance reasons or correctness reasons?  In other words, if I were to omit this function would the result change?

Reply:
It can effect when bottom values trigger, but that's not why it's here; it's here for performance reasons.  I have now pushed up a commit, 21403a88fa6c7bc5e75ca44be445c7235150f4f0 , that tries to clarify the comments concerning eta expansion.